library/core/src/ptr/const_ptr.rs

```diff
@@ -606,7 +606,9 @@ impl<T: ?Sized> *const T {
     /// Calculates the distance between two pointers. The returned value is in
     /// units of T: the distance in bytes divided by `mem::size_of::<T>()`.
     ///
-    /// This function is the inverse of [`offset`].
+    /// This function is the inverse of [`offset`]: it is valid to call if and only if
+    /// `self` could have been computed as `origin.offset(n)` for some `n`, and it will
+    /// then return that `n`.
```

Comment:
Using the phrase "if and only if" makes me wonder "maybe this has a more natural way to say it?", as it then goes on to express itself as a counterfactual. It's a dual, and the text is _correct_, but not everyone will find the mental motion you are suggesting a very natural one to make.

Is the idea that this is a "shortcut" and that if they grok this then they can just skip the rest of the text, here?

Reply:
I guess I'm too much of a logician, "if and only if" sounds natural to me. ;)

It's not a counterfactual, not in a logical sense anyway... the symbolic way of expressing this is
```
self.offset_from(origin) == Ok(n) <-> origin.offset(n) == Ok(self)
```
(where the `Ok` indicates non-UB execution)

Maybe there are better ways to put this into English?
"This function is well-defined with return value `n` if and only if `origin.offset(n)` is well-defined with return value `self`."

> Is the idea that this is a "shortcut" and that if they grok this then they can just skip the rest of the text, here?

Hm, not entirely sure. The claim about the inverse already existed before, but I thought it needs clarification since these are 2-input partial functions, so there is no obvious default meaning of "inverse".

Reply:
I find that slightly ambiguous since "if one would be UB to call then both are UB to call" here is under the scope of "the function's returned `isize`" but that already assumes that it was not UB to call. (A function that's UB doesn't return an `isize`. It's like the function returns `Result<isize, UB>`.)

Reply:
Oh, I'm certainly not trying to be precious about the exact wording. I just think that it should state this in two steps, because they're actually disjoint from the perspective of the programmer:
- The resulting value can be used as an argument in one function to produce the other one.
- The validity constraints (is that the right term?) for invoking either function are effectively mutual.

Reply:
The thing is, these steps are not separate though. It's a single equation:
```
self.offset_from(origin) == Ok(n) <-> origin.offset(n) == Ok(self)
```
This cannot be factored into two equations. To even talk about `origin.offset(n)` you need to know `n` but you only get that after assuming that `offset_from` is not UB, and vice versa.

Reply:
It's one equation but the sequent looks like
A, B ⊢ C
doesn't it?

Reply:
No, it's `A ⊣⊢ B`. It's completely symmetric.

Reply:
We can split the two directions of course:
- If `self.offset_from(origin)` is well-defined with result `n`, then `origin.offset(n)` is well-defined with result `self`.
- If `origin.offset(n)` is well-defined with result `self`, then `self.offset_from(origin)` is well-defined with result `n`.

But I don't see how that helps.

Reply:
Following https://github.com/rust-lang/rust/pull/95851 I have removed the note about the "inverse" entirely, and replaced it by
```
/// This is equivalent to `(self as isize - origin as isize) / (mem::size_of::<T>() as isize)`,
/// except that it has a lot more opportunities for UB, in exchange for the compiler
/// better understanding what you are doing.
```